README
NAME
    `Future' - represent an operation awaiting completion

SYNOPSIS
     my $future = Future->new;

     perform_some_operation(
        on_complete => sub {
           $future->done( @_ );
        }
     );

     $future->on_ready( sub {
        say "The operation is complete";
     } );

DESCRIPTION
    A `Future' object represents an operation that is currently in progress,
    or has recently completed. It can be used in a variety of ways to manage
    the flow of control, and data, through an asynchronous program.

    Some futures represent a single operation and are explicitly marked as
    ready by calling the `done' or `fail' methods. These are called "leaf"
    futures here, and are returned by the `new' constructor.

    Other futures represent a collection of sub-tasks, and are implicitly
    marked as ready depending on the readiness of their component futures as
    required. These are called "convergent" futures here as they converge
    control and data-flow back into one place. These are the ones returned
    by the various `wait_*' and `need_*' constructors.

    It is intended that library functions that perform asynchronous
    operations would use future objects to represent outstanding operations,
    and allow their calling programs to control or wait for these operations
    to complete. The implementation and the user of such an interface would
    typically make use of different methods on the class. The methods below
    are documented in two sections; those of interest to each side of the
    interface.

    It should be noted however, that this module does not in any way provide
    an actual mechanism for performing this asynchronous activity; it merely
    provides a way to create objects that can be used for control and data
    flow around those operations. It allows such code to be written in a
    neater, forward-reading manner, and simplifies many common patterns that
    are often involved in such situations.

    See also Future::Utils which contains useful loop-constructing
    functions, to run a future-returning function repeatedly in a loop.

  SUBCLASSING
    This class easily supports being subclassed to provide extra behavior,
    such as giving the `get' method the ability to block and wait for
    completion. This may be useful to provide `Future' subclasses with event
    systems, or similar.

    Each method that returns a new future object will use the invocant to
    construct its return value. If the constructor needs to perform
    per-instance setup it can override the `new' method, and take context
    from the given instance.

     sub new
     {
        my $proto = shift;
        my $self = $proto->SUPER::new;

        if( ref $proto ) {
           # Prototype was an instance
        }
        else {
           # Prototype was a class
        }

        return $self;
     }

    If an instance provides a method called `await', this will be called by
    the `get' and `failure' methods if the instance is pending.

     $f->await

    In most cases this should allow future-returning modules to be used as
    if they were blocking call/return-style modules, by simply appending a
    `get' call to the function or method calls.

     my ( $results, $here ) = future_returning_function( @args )->get;

    The examples directory in the distribution contains some examples of how
    futures might be integrated with various event systems.

  MODULE DOCUMENTATION
    Modules that provide future-returning functions or methods may wish to
    adopt the following styles in some way, to document the eventual return
    values from these futures.

     func( ARGS, HERE... ) ==> ( RETURN, VALUES... )

     OBJ->method( ARGS, HERE... ) ==> ( RETURN, VALUES... )

    Code returning a future that yields no values on success can use empty
    parentheses.

     func( ... ) ==> ()

  DEBUGGING
    By the time a `Future' object is destroyed, it ought to have been
    completed or cancelled. By enabling debug tracing of objects, this fact
    can be checked. If a future object is destroyed without having been
    completed or cancelled, a warning message is printed.

    This feature is enabled by setting an environment variable called
    `PERL_FUTURE_DEBUG' to some true value.

     $ PERL_FUTURE_DEBUG=1 perl -MFuture -E 'my $f = Future->new'
     Future=HASH(0xaa61f8) was constructed at -e line 1 and was lost near -e line 0 before it was ready.

    Note that due to a limitation of perl's `caller' function within a
    `DESTROY' destructor method, the exact location of the leak cannot be
    accurately determined. Often the leak will occur due to falling out of
    scope by returning from a function; in this case the leak location may
    be reported as being the line following the line calling that function.

     $ PERL_FUTURE_DEBUG=1 perl -MFuture
     sub foo {
        my $f = Future->new;
     }

     foo();
     print "Finished\n";

     Future=HASH(0x14a2220) was constructed at - line 2 and was lost near - line 6 before it was ready.
     Finished

CONSTRUCTORS
  $future = Future->new
  $future = $orig->new
    Returns a new `Future' instance to represent a leaf future. It will be
    marked as ready by any of the `done', `fail', or `cancel' methods. It
    can be called either as a class method, or as an instance method. Called
    on an instance it will construct another in the same class, and is
    useful for subclassing.

    This constructor would primarily be used by implementations of
    asynchronous interfaces.

  $future = Future->done( @values )
  $future = Future->fail( $exception, @details )
    Shortcut wrappers around creating a new `Future' then immediately
    marking it as done or failed.

  $future = Future->wrap( @values )
    If given a single argument which is already a `Future' reference, this
    will be returned unmodified. Otherwise, returns a new `Future' instance
    that is already complete, and will yield the given values.

    This will ensure that an incoming argument is definitely a `Future', and
    may be useful in such cases as adapting synchronous code to fit
    asynchronous libraries driven by `Future'.

  $future = Future->call( \&code, @args )
    A convenient wrapper for calling a `CODE' reference that is expected to
    return a future. In normal circumstances is equivalent to

     $future = $code->( @args )

    except that if the code throws an exception, it is wrapped in a new
    immediate fail future. If the return value from the code is not a
    blessed `Future' reference, an immediate fail future is returned instead
    to complain about this fact.

IMPLEMENTATION METHODS
    These methods would primarily be used by implementations of asynchronous
    interfaces.

  $future->done( @result )
    Marks that the leaf future is now ready, and provides a list of values
    as a result. (The empty list is allowed, and still indicates the future
    as ready). Cannot be called on a convergent future.

    If the future is already cancelled, this request is ignored. If the
    future is already complete with a result or a failure, an exception is
    thrown.

  Future->done( @result )
    May also be called as a class method, where it will construct a new
    Future and immediately mark it as done.

    Returns the `$future' to allow easy chaining to create an immediate
    future by

     return Future->done( ... )

  $code = $future->done_cb
    Returns a `CODE' reference that, when invoked, calls the `done' method.
    This makes it simple to pass as a callback function to other code.

    The same effect can be achieved using curry:

     $code = $future->curry::done;

  $future->fail( $exception, @details )
    Marks that the leaf future has failed, and provides an exception value.
    This exception will be thrown by the `get' method if called.

    The exception must evaluate as a true value; false exceptions are not
    allowed. Further details may be provided that will be returned by the
    `failure' method in list context. These details will not be part of the
    exception string raised by `get'.

    If the future is already cancelled, this request is ignored. If the
    future is already complete with a result or a failure, an exception is
    thrown.

  Future->fail( $exception, @details )
    May also be called as a class method, where it will construct a new
    Future and immediately mark it as failed.

    Returns the `$future' to allow easy chaining to create an immediate
    failed future by

     return Future->fail( ... )

  $code = $future->fail_cb
    Returns a `CODE' reference that, when invoked, calls the `fail' method.
    This makes it simple to pass as a callback function to other code.

    The same effect can be achieved using curry:

     $code = $future->curry::fail;

  $future->die( $message, @details )
    A convenient wrapper around `fail'. If the exception is a non-reference
    that does not end in a linefeed, its value will be extended by the file
    and line number of the caller, similar to the logic that `die' uses.

    Returns the `$future'.

  $future->on_cancel( $code )
    If the future is not yet ready, adds a callback to be invoked if the
    future is cancelled by the `cancel' method. If the future is already
    ready, throws an exception.

    If the future is cancelled, the callbacks will be invoked in the reverse
    order to that in which they were registered.

     $on_cancel->( $future )

  $future->on_cancel( $f )
    If passed another `Future' instance, the passed instance will be
    cancelled when the original future is cancelled. This method does
    nothing if the future is already complete.

  $cancelled = $future->is_cancelled
    Returns true if the future has been cancelled by `cancel'.

USER METHODS
    These methods would primarily be used by users of asynchronous
    interfaces, on objects returned by such an interface.

  $ready = $future->is_ready
    Returns true on a leaf future if a result has been provided to the
    `done' method, failed using the `fail' method, or cancelled using the
    `cancel' method.

    Returns true on a convergent future if it is ready to yield a result,
    depending on its component futures.

  $future->on_ready( $code )
    If the future is not yet ready, adds a callback to be invoked when the
    future is ready. If the future is already ready, invokes it immediately.

    In either case, the callback will be passed the future object itself.
    The invoked code can then obtain the list of results by calling the
    `get' method.

     $on_ready->( $future )

    Returns the `$future'.

  $future->on_ready( $f )
    If passed another `Future' instance, the passed instance will have its
    `done', `fail' or `cancel' methods invoked when the original future
    completes successfully, fails, or is cancelled respectively.

  $done = $future->is_done
    Returns true on a future if it is ready and completed successfully.
    Returns false if it is still pending, failed, or was cancelled.

  @result = $future->get
  $result = $future->get
    If the future is ready and completed successfully, returns the list of
    results that had earlier been given to the `done' method on a leaf
    future, or the list of component futures it was waiting for on a
    convergent future. In scalar context it returns just the first result
    value.

    If the future is ready but failed, this method raises as an exception
    the failure string or object that was given to the `fail' method.

    If the future was cancelled an exception is thrown.

    If it is not yet ready and is not of a subclass that provides an `await'
    method an exception is thrown. If it is subclassed to provide an `await'
    method then this is used to wait for the future to be ready, before
    returning the result or propagating its failure exception.

  @values = Future->unwrap( @values )
    If given a single argument which is a `Future' reference, this method
    will call `get' on it and return the result. Otherwise, it returns the
    list of values directly in list context, or the first value in scalar.
    Since it involves an implicit `await', this method can only be used on
    immediate futures or subclasses that implement `await'.

    This will ensure that an outgoing argument is definitely not a `Future',
    and may be useful in such cases as adapting synchronous code to fit
    asynchronous libraries that return `Future' instances.

  $future->on_done( $code )
    If the future is not yet ready, adds a callback to be invoked when the
    future is ready, if it completes successfully. If the future completed
    successfully, invokes it immediately. If it failed or was cancelled, it
    is not invoked at all.

    The callback will be passed the result passed to the `done' method.

     $on_done->( @result )

    Returns the `$future'.

  $future->on_done( $f )
    If passed another `Future' instance, the passed instance will have its
    `done' method invoked when the original future completes successfully.

  $failed = $future->is_failed
    Returns true on a future if it is ready and it failed. Returns false if
    it is still pending, completed successfully, or was cancelled.

  $exception = $future->failure
  $exception, @details = $future->failure
    Returns the exception passed to the `fail' method, `undef' if the future
    completed successfully via the `done' method, or raises an exception if
    called on a future that is not yet ready.

    If called in list context, will additionally yield a list of the details
    provided to the `fail' method.

    Because the exception value must be true, this can be used in a simple
    `if' statement:

     if( my $exception = $future->failure ) {
        ...
     }
     else {
        my @result = $future->get;
        ...
     }

  $future->on_fail( $code )
    If the future is not yet ready, adds a callback to be invoked when the
    future is ready, if it fails. If the future has already failed, invokes
    it immediately. If it completed successfully or was cancelled, it is not
    invoked at all.

    The callback will be passed the exception and details passed to the
    `fail' method.

     $on_fail->( $exception, @details )

    Returns the `$future'.

  $future->on_fail( $f )
    If passed another `Future' instance, the passed instance will have its
    `fail' method invoked when the original future fails.

    To invoke a `done' method on a future when another one fails, use a CODE
    reference:

     $future->on_fail( sub { $f->done( @_ ) } );

  $future->cancel
    Requests that the future be cancelled, immediately marking it as ready.
    This will invoke all of the code blocks registered by `on_cancel', in
    the reverse order. When called on a convergent future, all its component
    futures are also cancelled. It is not an error to attempt to cancel a
    future that is already complete or cancelled; it simply has no effect.

    Returns the `$future'.

  $code = $future->cancel_cb
    Returns a `CODE' reference that, when invoked, calls the `cancel'
    method. This makes it simple to pass as a callback function to other
    code.

    The same effect can be achieved using curry:

     $code = $future->curry::cancel;

SEQUENCING METHODS
    The following methods all return a new future to represent the
    combination of its invocant followed by another action given by a code
    reference. The combined activity waits for the first future to be ready,
    then may invoke the code depending on the success or failure of the
    first, or may run it regardless. The returned sequence future represents
    the entire combination of activity.

    In some cases the code should return a future; in some it should return
    an immediate result. If a future is returned, the combined future will
    then wait for the result of this second one. If the combinined future is
    cancelled, it will cancel either the first future or the second,
    depending whether the first had completed. If the code block throws an
    exception instead of returning a value, the sequence future will fail
    with that exception as its message and no further values.

    As it is always a mistake to call these sequencing methods in void
    context and lose the reference to the returned future (because
    exception/error handling would be silently dropped), this method warns
    in void context.

  $future = $f1->then( \&done_code )
    Returns a new sequencing `Future' that runs the code if the first
    succeeds. Once `$f1' succeeds the code reference will be invoked and is
    passed the list of results. It should return a future, `$f2'. Once `$f2'
    completes the sequence future will then be marked as complete with
    whatever result `$f2' gave. If `$f1' fails then the sequence future will
    immediately fail with the same failure and the code will not be invoked.

     $f2 = $done_code->( @result )

  $future = $f1->else( \&fail_code )
    Returns a new sequencing `Future' that runs the code if the first fails.
    Once `$f1' fails the code reference will be invoked and is passed the
    failure and details. It should return a future, `$f2'. Once `$f2'
    completes the sequence future will then be marked as complete with
    whatever result `$f2' gave. If `$f1' succeeds then the sequence future
    will immediately succeed with the same result and the code will not be
    invoked.

     $f2 = $fail_code->( $exception, @details )

  $future = $f1->then( \&done_code, \&fail_code )
    The `then' method can also be passed the `$fail_code' block as well,
    giving a combination of `then' and `else' behaviour.

    This operation is designed to be compatible with the semantics of other
    future systems, such as Javascript's Q or Promises/A libraries.

  $future = $f1->transform( %args )
    Returns a new sequencing `Future' that wraps the one given as `$f1'.
    With no arguments this will be a trivial wrapper; `$future' will
    complete or fail when `$f1' does, and `$f1' will be cancelled when
    `$future' is.

    By passing the following named arguments, the returned `$future' can be
    made to behave differently to `$f1':

    done => CODE
            Provides a function to use to modify the result of a successful
            completion. When `$f1' completes successfully, the result of its
            `get' method is passed into this function, and whatever it
            returns is passed to the `done' method of `$future'

    fail => CODE
            Provides a function to use to modify the result of a failure.
            When `$f1' fails, the result of its `failure' method is passed
            into this function, and whatever it returns is passed to the
            `fail' method of `$future'.

  $future = $f1->then_with_f( \&code )
    Returns a new sequencing `Future' that runs the code if the first
    succeeds. Identical to `then', except that the code reference will be
    passed both the original future, `$f1', and its result.

     $f2 = $code->( $f1, @result )

    This is useful for conditional execution cases where the code block may
    just return the same result of the original future. In this case it is
    more efficient to return the original future itself.

  $future = $f->then_done( @result )
  $future = $f->then_fail( $exception, @details )
    Convenient shortcuts to returning an immediate future from a `then'
    block, when the result is already known.

  $future = $f1->else_with_f( \&code )
    Returns a new sequencing `Future' that runs the code if the first fails.
    Identical to `else', except that the code reference will be passed both
    the original future, `$f1', and its exception and details.

     $f2 = $code->( $f1, $exception, @details )

    This is useful for conditional execution cases where the code block may
    just return the same result of the original future. In this case it is
    more efficient to return the original future itself.

  $future = $f->else_done( @result )
  $future = $f->else_fail( $exception, @details )
    Convenient shortcuts to returning an immediate future from a `else'
    block, when the result is already known.

  $future = $f1->followed_by( \&code )
    Returns a new sequencing `Future' that runs the code regardless of
    success or failure. Once `$f1' is ready the code reference will be
    invoked and is passed one argument, `$f1'. It should return a future,
    `$f2'. Once `$f2' completes the sequence future will then be marked as
    complete with whatever result `$f2' gave.

     $f2 = $code->( $f1 )

  $future = $f1->and_then( \&code )
    An older form of `then_with_f'; this method passes only the original
    future itself to the code, not its result. The code would have to call
    `get' on the future to obtain the result.

     $f2 = $code->( $f1 )

    This method will be removed in a later version; use `then_with_f' in new
    code. As a reminder, this method will now cause a warning every time it
    is invoked.

  $future = $f1->or_else( \&code )
    An older form of `else_with_f'; this method passes only the original
    future itself to the code, not its failure and details. The code would
    have to call `failure' on the future to obtain the result.

     $f2 = $code->( $f1 )

    This method will be removed in a later version; use `else_with_f' in new
    code. As a reminder, this method will now cause a warning every time it
    is invoked.

  $future = $f1->without_cancel
    Returns a new sequencing `Future' that will complete with the success or
    failure of the original future, but if cancelled, will not cancel the
    original. This may be useful if the original future represents an
    operation that is being shared among multiple sequences; cancelling one
    should not prevent the others from running too.

CONVERGENT FUTURES
    The following constructors all take a list of component futures, and
    return a new future whose readiness somehow depends on the readiness of
    those components. The first derived class component future will be used
    as the prototype for constructing the return value, so it respects
    subclassing correctly, or failing that a plain `Future'.

  $future = Future->wait_all( @subfutures )
    Returns a new `Future' instance that will indicate it is ready once all
    of the sub future objects given to it indicate that they are ready,
    either by success, failure or cancellation. Its result will a list of
    its component futures.

    When given an empty list this constructor returns a new immediately-done
    future.

    This constructor would primarily be used by users of asynchronous
    interfaces.

  $future = Future->wait_any( @subfutures )
    Returns a new `Future' instance that will indicate it is ready once any
    of the sub future objects given to it indicate that they are ready,
    either by success or failure. Any remaining component futures that are
    not yet ready will be cancelled. Its result will be the result of the
    first component future that was ready; either success or failure. Any
    component futures that are cancelled are ignored, apart from the final
    component left; at which point the result will be a failure.

    When given an empty list this constructor returns an immediately-failed
    future.

    This constructor would primarily be used by users of asynchronous
    interfaces.

  $future = Future->needs_all( @subfutures )
    Returns a new `Future' instance that will indicate it is ready once all
    of the sub future objects given to it indicate that they have completed
    successfully, or when any of them indicates that they have failed. If
    any sub future fails, then this will fail immediately, and the remaining
    subs not yet ready will be cancelled. Any component futures that are
    cancelled will cause an immediate failure of the result.

    If successful, its result will be a concatenated list of the results of
    all its component futures, in corresponding order. If it fails, its
    failure will be that of the first component future that failed. To
    access each component future's results individually, use `done_futures'.

    When given an empty list this constructor returns a new immediately-done
    future.

    This constructor would primarily be used by users of asynchronous
    interfaces.

  $future = Future->needs_any( @subfutures )
    Returns a new `Future' instance that will indicate it is ready once any
    of the sub future objects given to it indicate that they have completed
    successfully, or when all of them indicate that they have failed. If any
    sub future succeeds, then this will succeed immediately, and the
    remaining subs not yet ready will be cancelled. Any component futures
    that are cancelled are ignored, apart from the final component left; at
    which point the result will be a failure.

    If successful, its result will be that of the first component future
    that succeeded. If it fails, its failure will be that of the last
    component future to fail. To access the other failures, use
    `failed_futures'.

    Normally when this future completes successfully, only one of its
    component futures will be done. If it is constructed with multiple that
    are already done however, then all of these will be returned from
    `done_futures'. Users should be careful to still check all the results
    from `done_futures' in that case.

    When given an empty list this constructor returns an immediately-failed
    future.

    This constructor would primarily be used by users of asynchronous
    interfaces.

METHODS ON CONVERGENT FUTURES
    The following methods apply to convergent (i.e. non-leaf) futures, to
    access the component futures stored by it.

  @f = $future->pending_futures
  @f = $future->ready_futures
  @f = $future->done_futures
  @f = $future->failed_futures
  @f = $future->cancelled_futures
    Return a list of all the pending, ready, done, failed, or cancelled
    component futures. In scalar context, each will yield the number of such
    component futures.

TRACING METHODS
  $future = $future->set_label( $label )
  $label = $future->label
    Chaining mutator and accessor for the label of the `Future'. This should
    be a plain string value, whose value will be stored by the future
    instance for use in debugging messages or other tooling, or similar
    purposes.

  [ $sec, $usec ] = $future->btime
  [ $sec, $usec ] = $future->rtime
    Accessors that return the tracing timestamps from the instance. These
    give the time the instance was contructed ("birth" time, `btime') and
    the time the result was determined (the "ready" time, `rtime'). Each
    result is returned as a two-element ARRAY ref, containing the epoch time
    in seconds and microseconds, as given by `Time::HiRes::gettimeofday'.

    In order for these times to be captured, they have to be enabled by
    setting `$Future::TIMES' to a true value. This is initialised true at
    the time the module is loaded if either `PERL_FUTURE_DEBUG' or

Changes
Revision history for Future

0.30    2014/11/26 14:29:28
        [CHANGES]
         * Rename 'dependent' futures to 'convergent'
         * Removed examples/ scripts that now exist as independent modules
         * Added ->without_cancel
         * Sanity-check the $code argument to ->on_{ready,done,fail,cancel} to
           ensure it is callable or a Future
        
        [BUGFIXES]
         * Ensure that 'ready_at' is always set in DEBUG mode
         * Fix DEBUG 'lost_at' line number reporting tests for latest
           bleadperl (RT99002)
         * Ensure that if Future::Utils::repeat condition code dies, that is
           passed to the result Future and not propagated to the caller
           (RT100067)
         * Failure by returning a non-Future from a sequencing code block
           should report as a failed Future, not throw exception to caller

0.29    2014/07/17 12:18:12
        [CHANGES]
         * Added Test::Future
         * Stronger deprecations - repeat {} on failures warns every time,
           ->and_then / ->or_else warn once

        [BUGFIXES]
         * Define the behaviour of dependent futures when components are
           cancelled. (Partially fixes RT96685)
         * Use Module::Build->prompt (RT96409)
         * Ensure that repeat on an empty foreach list or empty generator
           without 'otherwise' behaves correctly, just yield an immediate

0.28    2014/06/08 22:43:40
        [CHANGES]
         * Added ->label
         * Added ->btime, rtime, elapsed tracing timers
         * Better handling of 'breaks' version detection

0.27    2014/06/06 17:42:27
        [BUGFIXES]
         * Depend on Carp 1.25 for the new message format with trailing
           period, so tests work

0.26    2014/06/01 12:52:53
        [CHANGES]
         * Added ->is_failed accessor
         * Implement ->export_to_level in Future::Utils
         * Print a warning about lost sequence Futures
         * Allow Future->done and Future->fail as simple class constructors
           to return immediates
         * Added Future->unwrap

        [BUGFIXES]
         * Ensure that sequence futures are weaken()ed in the forward
           direction.
           **NOTE** This will potentially break existing code that depended on
             strong references. This old code was, however, broken.

0.25    2014/02/22 03:47:08
        [BUGFIXES]
         * Fix warning-matching test in unit test for both older and newer
           versions of Carp

0.24    2014/02/21 17:57:49
        [CHANGES]
         * Have repeat print a warning if it is asked to retry over a failure
         * Change documentation to suggest try_repeat instead of repeat for
           retries over failure
         * Check at call time that sequencing callbacks really are callable,
           leading to neater error messages (RT93164)

0.23    2014/01/19 15:26:55
        [CHANGES]
         * Link to Futures advent calendar 2013
         * Fixes/additions to Phrasebook documentation, including section
           about tree recursion

        [BUGFIXES]
         * Ensure that late addition of additional items to a fmap foreach
           array works correctly even with concurrency

0.22    2014/01/12 03:12:18
        [CHANGES]
         * Ignore ->done or ->fail on cancelled Futures
         * Added ->then_done, ->then_fail, ->else_done, ->else_fail
         * Neaten up naming of fmap* family - provide both long and short
           names for each function
         * Added Future::Utils::call and call_with_escape
         * Ensure that dependent futures on subclasses tries to use derived
           futures as prototype if possible

0.21    2013/12/29 18:14:41
        [CHANGES]
         * Major performance improvement by folding out some layers of sub {}
           wrapping in sequencing operations
         * Added ->then_with_f and ->else_with_f
        
        [BUGFIXES]
         * Don't start another trial after cancelling a repeat() (RT91147)

0.20    2013/11/18 19:14:27
        [CHANGES]
         * Include an indication of done/failed/cancelled status of a Future
           when ->done or ->failing an already-ready one

        [BUGFIXES]
         * Declare requires perl 5.8 because it fails on 5.6 smokers - no idea
           why
         * Fix a couple of typoes in docs (RT89185)

0.19    2013/09/27 13:31:16
        [BUGFIXES]
         * Guard against odd things happening during ->cancel at global
           destruction (RT88967)

0.18    2013/09/20 19:09:57
        [CHANGES]
         * Added 'try_repeat' and 'try_repeat_until_success' aliases
         * @CARP_NOT trust between Future and Future::Utils

        [BUGFIXES]
         * Fix to concurrent non-immediate + immediate fmap* return values

0.17    2013/09/07 16:53:47
        [CHANGES]
         * Performance improvement by using direct member access instead of
           accessor methods
         * Documentation updates; suggestion of documentation style for
           Future-returning code

        [BUGFIXES]
         * Respect subclassing of immediate->followed_by and dependent futures
           with mixed subclass or immediate components

0.16    CHANGES:
         * Proper behaviour of ->wait_all and ->needs_all on an empty list -
           just return empty immediate done
         * Proper behaviour of ->wait_any and ->needs_any on an empty list -
           return an immediate failure
         * Performance improvement to ->done for immediate Future->new->done
         * Keep a count of pending child futures to avoid quadratic behaviour
           due to linear scan of children every time one completes
         * Improve efficiency of Future::Utils::repeat and fmap* when trials
           return immediates
         * Make repeat and fmap* 'return' argument optional by cloning the
           first non-immediate trial
         * Rework unit tests to avoid dependency on Test::Warn

0.15    CHANGES:
         * Added Future->call constructor

Build.PL
use strict;
use warnings;

use Module::Build;

# This version of Future contains an important bugfix around weak references
# in sequence Futures. Unfortunately, a lot of existing CPAN code is known to
# rely on this behaviour, and will break if this module is upgraded.
#
# Abort if any of the following modules are installed at versions less than
# the first known-working version. They must be updated before Future can be
# installed.
my %FIXED = (
   'IO::Async' => '0.62',
   'IO::Async::SSL' => '0.14',
   'Net::Async::CassandraCQL' => '0.11',
   'Net::Async::FTP' => '0.08',
   'Net::Async::HTTP' => '0.34',
   'Net::Async::WebSocket' => '0.08',
);

my $printed;
foreach my $module ( sort keys %FIXED ) {
   my $needsver = $FIXED{$module};
   ( my $modfile = "$module.pm" ) =~ s{::}{/}g;

   next unless eval { require $modfile };
   next if ( my $instver = $module->VERSION ) >= $needsver;

   print STDERR "Installing this version of Future will fix a bug that the following installed\n".
                "modules rely on. You must upgrade these modules to a later version after\n".
                "Future is installed, or they will not work correctly.\n\n" unless $printed;

   print STDERR "  * $module (installed $instver; need at least $needsver)\n";
   $printed++;
}
print STDERR "\n" if $printed;

if( $printed and -t STDIN ) {
   # Attended update; might as well ask the user to confirm and exit if not
   my $reply = Module::Build->prompt(
      "Are you still sure you wish to go ahead with this upgrade?\n" .
         "[enter 'yes' to continue]: ",
      "no"
   );

   die "Aborting install due to broken dependent modules\n" unless $reply =~ m/^y/i;
}

my $build = Module::Build->new(
   module_name => 'Future',
   build_requires => {
      'Test::Identity' => 0,
      'Test::Fatal' => 0,
      'Test::More' => '0.88', # done_testing
      'Test::Refcount' => 0,
   },
   requires => {
      'perl' => '5.008', # fails on 5.6 smokers; no idea why
      'Carp' => '1.25', # new message format with trailing period
      'Test::Builder::Module' => 0,
      'Time::HiRes' => 0,
   },
   meta_merge => {
      # It's unlikely at the time of writing that any CPAN client actually
      # pays attention to this field, but it's nice to declare it on CPAN
      # anyway so people will know I want to use it; maybe one day clients
      # will follow it...
      x_breaks => { do {
         map { $_ => "< $FIXED{$_}" } keys %FIXED
      }},
   },
   auto_configure_requires => 0, # Don't add M::B to configure_requires
   license => 'perl',
   create_makefile_pl => 'traditional',
   create_license => 1,
   create_readme  => 1,
   meta_merge => {
      resources => {
         x_IRC => "irc://irc.perl.org/#io-async",
      },
   },
);

$build->create_build_script;
